// @moneybutton/paymail-client/dist/paymail-client.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var moment = _interopDefault(require('moment'));
var fetch = _interopDefault(require('isomorphic-fetch'));

const Capabilities = {
  pki: 'pki',
  paymentDestination: 'paymentDestination',
  senderValidation: '6745385c3fc0',
  verifyPublicKeyOwner: 'a9f510c16bde',
  publicProfile: 'f12f968c92d6',
  receiveTransaction: 'receive-transactions-alpha-state'
};

// import { DnsOverHttps } from "./dns-over-https"
class DnsClient {
  constructor(dns, doh) {
    this.dns = dns;
    this.doh = doh;
  }

  async checkSrv(aDomain) {
    return new Promise((resolve, reject) => {
      this.dns.resolveSrv(`_bsvalias._tcp.${aDomain}`, async (err, result) => {
        try {
          if (err && (err.code === 'ENODATA' || err.code === 'ENOTFOUND')) {
            return resolve({
              domain: aDomain,
              port: 443,
              isSecure: true
            });
          }

          if (err) {
            return reject(err);
          }

          const {
            name: domainFromDns,
            port,
            isSecure
          } = result[0];
          resolve({
            domain: domainFromDns,
            port,
            isSecure: this.checkDomainIsSecure(domainFromDns, aDomain) || isSecure
          });
        } catch (err) {
          return reject(err);
        }
      });
    }).then(result => {
      if (result.isSecure) {
        return result;
      } else {
        return this.validateDnssec(aDomain);
      }
    }, err => {
      console.error(err);
      return err;
    });
  }

  checkDomainIsSecure(srvResponseDomain, originalDomain) {
    if (this.domainsAreEqual(srvResponseDomain, originalDomain)) {
      return true;
    } else if (this.responseIsWwwSubdomain(srvResponseDomain, originalDomain)) {
      return true;
    } else if (this.isHandcashDomain(originalDomain)) {
      // tell rafa to fix handcash and we can remove the special case :)
      return this.domainsAreEqual('handcash-paymail-production.herokuapp.com', srvResponseDomain) || this.domainsAreEqual('handcash-cloud-production.herokuapp.com', srvResponseDomain);
    } else if (this.isHandcashInternalDomain(originalDomain)) {
      return this.domainsAreEqual('handcash-cloud-staging.herokuapp.com', srvResponseDomain);
    } else if (this.domainsAreEqual('localhost', srvResponseDomain)) {
      return true;
    } else if (this.isMoneyButtonDomain(srvResponseDomain)) {
      return true;
    } else {
      return false;
    }
  }

  isMoneyButtonDomain(aDomain) {
    return this.domainsAreEqual(aDomain, 'moneybutton.com') || this.domainsAreEqual(aDomain, 'www.moneybutton.com');
  }

  responseIsWwwSubdomain(srvResponseDomain, originalDomain) {
    return this.domainsAreEqual(srvResponseDomain, `www.${originalDomain}`);
  }

  isHandcashDomain(aDomain) {
    return this.domainsAreEqual('handcash.io', aDomain);
  }

  isHandcashInternalDomain(aDomain) {
    return this.domainsAreEqual('internal.handcash.io', aDomain);
  }

  async validateDnssec(aDomain) {
    const dnsResponse = await this.doh.queryBsvaliasDomain(aDomain);

    if (dnsResponse.Status !== 0 || !dnsResponse.Answer) {
      throw new Error(`Insecure domain.`);
    }

    const data = dnsResponse.Answer[0].data.split(' ');
    const port = data[2];
    const responseDomain = data[3];

    if (!dnsResponse.AD && !this.domainsAreEqual(aDomain, responseDomain)) {
      throw new Error(`Insecure domain.`);
    }

    return {
      port,
      domain: responseDomain,
      isSecure: dnsResponse.AD
    };
  }

  domainsAreEqual(domain1, domain2) {
    return domain1.replace(/\.$/, '') === domain2.replace(/\.$/, '');
  }

}

class DnsOverHttps {
  constructor(fetch$$1, config) {
    this.fetch = fetch$$1;
    this.config = config;
  }

  async resolveSrv(aDomain) {
    const response = await this.fetch(`${this.config.baseUrl}?name=${aDomain}&type=SRV&cd=0`);
    const body = await response.json();
    return body;
  }

  async queryBsvaliasDomain(aDomain) {
    return this.resolveSrv(`_bsvalias._tcp.${aDomain}`);
  }

}

class EndpointResolver {
  constructor(dns = null, fetch$$1) {
    this.dnsClient = new DnsClient(dns, new DnsOverHttps(fetch$$1, {
      baseUrl: 'https://dns.google.com/resolve'
    }));
    this.fetch = fetch$$1;
    this._cache = {};
  }

  static create(dnsClient, fetch$$1) {
    const instance = new EndpointResolver(null, fetch$$1);
    instance.dnsClient = dnsClient;
    return instance;
  }

  async getIdentityUrlFor(aPaymail) {
    const [alias, domain] = aPaymail.split('@');
    await this.ensureCapabilityFor(domain, Capabilities.pki);
    const apiDescriptor = await this.getApiDescriptionFor(domain);
    const identityUrl = apiDescriptor.capabilities.pki.replace('{alias}', alias).replace('{domain.tld}', domain);
    return identityUrl;
  }

  async getAddressUrlFor(aPaymail) {
    const [alias, domain] = aPaymail.split('@');
    await this.ensureCapabilityFor(domain, Capabilities.paymentDestination);
    const apiDescriptor = await this.getApiDescriptionFor(domain);
    const addressUrl = apiDescriptor.capabilities.paymentDestination.replace('{alias}', alias).replace('{domain.tld}', domain);
    return addressUrl;
  }

  async getVerifyUrlFor(aPaymail, aPubkey) {
    const [alias, domain] = aPaymail.split('@');
    await this.ensureCapabilityFor(domain, Capabilities.verifyPublicKeyOwner);
    const apiDescriptor = await this.getApiDescriptionFor(domain);
    const url = apiDescriptor.capabilities[Capabilities.verifyPublicKeyOwner].replace('{alias}', alias).replace('{domain.tld}', domain).replace('{pubkey}', aPubkey);
    return url;
  }

  async getPublicProfileUrlFor(aPaymail) {
    const [alias, domain] = aPaymail.split('@');
    await this.ensureCapabilityFor(domain, Capabilities.publicProfile);
    const apiDescriptor = await this.getApiDescriptionFor(domain);
    const url = apiDescriptor.capabilities[Capabilities.publicProfile].replace('{alias}', alias).replace('{domain.tld}', domain);
    return url;
  }

  async getSendTxUrlFor(aPaymail) {
    const [alias, domain] = aPaymail.split('@');
    await this.ensureCapabilityFor(domain, Capabilities.receiveTransaction);
    const apiDescriptor = await this.getApiDescriptionFor(domain);
    const url = apiDescriptor.capabilities[Capabilities.receiveTransaction].replace('{alias}', alias).replace('{domain.tld}', domain);
    return url;
  }

  async domainHasCapability(aDomain, capability) {
    const apiDescriptor = await this.getApiDescriptionFor(aDomain);
    return !!apiDescriptor.capabilities[capability];
  }

  async getApiDescriptionFor(aDomain) {
    if (this._cache[aDomain]) {
      return this._cache[aDomain];
    }

    const {
      domain,
      port
    } = await this.getWellKnownBaseUrl(aDomain);
    const apiDescriptor = this.fetchApiDescriptor(domain, port);
    this._cache[aDomain] = apiDescriptor;
    return apiDescriptor;
  }

  async fetchApiDescriptor(domain, port) {
    const protocol = domain === 'localhost' || domain === 'localhost.' ? 'http' : 'https';
    const wellKnown = await this.fetch(`${protocol}://${domain}:${port}/.well-known/bsvalias`, {
      credentials: 'omit'
    });
    const apiDescriptor = await wellKnown.json();
    return apiDescriptor;
  }

  async getWellKnownBaseUrl(aDomain) {
    return this.dnsClient.checkSrv(aDomain);
  }

  async ensureCapabilityFor(aDomain, aCapability) {
    if (!(await this.domainHasCapability(aDomain, aCapability))) {
      throw new Error(`Unknown capability "${aCapability}" for "${aDomain}"`);
    }
  }

}

class VerifiableMessage {
  constructor(parts, bsv = null) {
    if (bsv === null) {
      bsv = require('bsv');
      bsv.Message = require('bsv/message');
    }

    this.bsv = bsv;
    const concatenated = Buffer.from(parts.join(''));
    const hashed = this.bsv.crypto.Hash.sha256(concatenated).toString('hex');
    this.message = new this.bsv.Message(hashed);
  }

  static forBasicAddressResolution({
    senderHandle,
    amount,
    dt,
    purpose
  }) {
    if (dt.toISOString) {
      dt = dt.toISOString();
    }

    return new VerifiableMessage([senderHandle, amount || '0', dt, purpose]);
  }

  sign(wifPrivateKey) {
    return this.message.sign(this.bsv.PrivateKey.fromWIF(wifPrivateKey));
  }

  verify(keyAddress, signature) {
    return this.message.verify(keyAddress, signature);
  }

}

class RequestBodyFactory {
  constructor(clock) {
    this.clock = clock;
  }

  buildBodyToRequestAddress(senderInfo, privateKey = null) {
    const {
      senderHandle,
      amount,
      senderName,
      purpose,
      pubkey,
      signature: providedSignature
    } = senderInfo;

    if (!providedSignature && privateKey === null) {
      throw new Error('Missing private key or signature');
    }

    let dt, signature;

    if (providedSignature) {
      if (!senderInfo.dt) {
        throw new Error('missing datetime for given signature');
      }

      dt = senderInfo.dt;
      signature = providedSignature;
    } else {
      dt = this.clock.now();
      signature = VerifiableMessage.forBasicAddressResolution({
        senderHandle,
        amount,
        dt,
        purpose
      }).sign(privateKey);
    }

    return {
      senderHandle,
      senderName,
      purpose,
      dt,
      amount: amount || null,
      pubkey,
      signature
    };
  }

  buildBodySendTx(transactions, metadata, reference) {
    return {
      transactions,
      metadata,
      reference
    };
  }

}

class Clock {
  now() {
    return moment();
  }

}

class PaymailNotFound extends Error {
  constructor(message, paymail) {
    super(message);
    this.paymail = paymail;
  }

}

class BrowserDns {
  constructor(fetch$$1) {
    this.doh = new DnsOverHttps(fetch$$1, {
      baseUrl: 'https://dns.google.com/resolve'
    });
  }

  async resolveSrv(aDomain, aCallback) {
    try {
      const response = await this.doh.resolveSrv(aDomain);

      if (response.Status === 0) {
        const data = response.Answer.map(record => {
          const [priority, weight, port, name] = record.data.split(' ');
          return {
            priority,
            weight,
            port,
            name,
            isSecure: response.AD
          };
        });
        aCallback(null, data);
      } else {
        aCallback(new Error('error during dns query'));
      }
    } catch (e) {
      aCallback(e);
    }
  }

}

class PaymailClient {
  constructor(dns = null, fetch2 = null, clock = null, bsv = null) {
    if (fetch2 === null) {
      fetch2 = fetch;
    }

    if (dns === null) {
      dns = new BrowserDns(fetch2);
    }

    if (bsv === null) {
      bsv = require('bsv');
    }

    this.bsv = bsv;
    this.resolver = new EndpointResolver(dns, fetch2);
    this.fetch = fetch2;
    this.requestBodyFactory = new RequestBodyFactory(clock !== null ? clock : new Clock());
  }
  /**
   * Uses pki flow to query for an identity key for a given paymail address.
   *
   * @param {String} paymail - a paymail address
   */


  async getPublicKey(paymail) {
    const identityUrl = await this.resolver.getIdentityUrlFor(paymail);
    const response = await this.fetch(identityUrl, {
      credentials: 'omit'
    });
    const {
      pubkey
    } = await response.json();
    return pubkey;
  }
  /**
   * Uses `Basic Address Resolution` flow to query for a payment for output for the
   * given paymail address.
   *
   * @param {String} aPaymail - a paymail address
   * @param {Object} senderInfo - Object containing sender info
   * @param {String} senderInfo.senderHandle - Sender paymail address
   * @param {String} senderInfo.amount - Optional. Required amount.
   * @param {String} senderInfo.senderName - Optional. Sender name.
   * @param {String} senderInfo.purpose - Optional. Purpose of the payment.
   * @param {String} senderInfo.pubkey - Optional. Public key used to sign the message.
   * @param {String} senderInfo.signature - Optional. Valid signature according to paymail specification.
   * @param {String} privateKey - Optional. private key to sign the request.
   */


  async getOutputFor(aPaymail, senderInfo, privateKey = null) {
    const addressUrl = await this.resolver.getAddressUrlFor(aPaymail);
    const response = await this.fetch(addressUrl, {
      method: 'POST',
      credentials: 'omit',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(this.requestBodyFactory.buildBodyToRequestAddress(senderInfo, privateKey))
    });

    if (!response.ok) {
      throw new PaymailNotFound(`Paymail not found: ${aPaymail}`, aPaymail);
    }

    const {
      output
    } = await response.json();
    return output;
  }
  /**
   * Verify if the given public address belongs to the given
   * paymail address.
   *
   * @param {String} pubkey - Public key to check.
   * @param {String} paymail - a paymail address
   */


  async verifyPubkeyOwner(pubkey, paymail) {
    const url = await this.resolver.getVerifyUrlFor(paymail, pubkey);
    const response = await this.fetch(url, {
      credentials: 'omit'
    });
    const body = await response.json();
    const {
      match
    } = body;
    return match;
  }
  /**
   * Verifies if a given signature is valid for a given message. It uses
   * different strategies depending on the capabilities of the server
   * and the parameters Given. The priority order is.
   * - If paymail is not provided, then normal signature verification is performed.
   * - Use provided key (and check that belongs to given paymail address).
   * - Get a new pubkey for given paymail address using pki.
   * - If there is no way to intereact with the owner of the domain to verify the public key it returns false.
   *
   * @param {Message} message - Message to verify
   * @param {String} signature - Signature
   * @param {String} paymail - Signature owner paymail
   * @param {String} pubkey - Optional. Public key that validates the signature.
   */


  async isValidSignature(message, signature, paymail = null, pubkey = null) {
    if (paymail == null && pubkey === null) {
      throw new Error('Must specify either paymail or pubkey');
    }

    let senderPublicKey;

    if (paymail) {
      if (pubkey && (await this.resolver.domainHasCapability(paymail.split('@')[1], Capabilities.verifyPublicKeyOwner))) {
        if (await this.verifyPubkeyOwner(pubkey, paymail)) {
          senderPublicKey = this.bsv.PublicKey.fromString(pubkey);
        } else {
          return false;
        }
      } else {
        const hasPki = await this.resolver.domainHasCapability(paymail.split('@')[1], Capabilities.pki);

        if (hasPki) {
          const identityKey = await this.getPublicKey(paymail);
          senderPublicKey = this.bsv.PublicKey.fromString(identityKey);
        } else {
          return false;
        }
      }
    }

    const senderKeyAddress = this.bsv.Address.fromPublicKey(senderPublicKey || pubkey);

    try {
      const verified = message.verify(senderKeyAddress.toString(), signature);
      return verified;
    } catch (err) {
      return false;
    }
  }
  /**
   * Gets the public profile information using the "Public Profile" protocol.
   *
   * @param {String} paymail - a paymail address
   * @param {String} s - the preferred size of the image
   */


  async getPublicProfile(paymail) {
    let publicProfileUrl = await this.resolver.getPublicProfileUrlFor(paymail);
    const response = await this.fetch(publicProfileUrl, {
      credentials: 'omit'
    });

    if (!response.ok) {
      const body = await response.json();
      throw new Error(`Server failed with: ${JSON.stringify(body)}`);
    }

    const {
      avatar,
      name
    } = await response.json();
    return {
      avatar,
      name
    };
  }

  async sendRawTx(targetPaymail, transactions = [], metadata = {}, reference = null) {
    if (!transactions.every(tx => tx.hex)) {
      throw new Error('Transactions should include "hex" field');
    }

    if (transactions.length === 0) {
      throw new Error('Transaction array should not be empty.');
    }

    let receiveTxUrl = await this.resolver.getSendTxUrlFor(targetPaymail);
    const response = await this.fetch(receiveTxUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(this.requestBodyFactory.buildBodySendTx(transactions, metadata, reference))
    });

    if (!response.ok) {
      const body = await response.json();
      throw new Error(`Server failed with: ${JSON.stringify(body)}`);
    }

    return response.json();
  }

}

exports.PaymailClient = PaymailClient;
exports.VerifiableMessage = VerifiableMessage;
exports.RequestBodyFactory = RequestBodyFactory;
exports.Clock = Clock;
exports.PaymailNotFound = PaymailNotFound;
exports.BrowserDns = BrowserDns;
