// @moneybutton/express-paymail/lib/buildGetAddressRouter.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildGetAddressRouter = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _express = _interopRequireDefault(require("express"));

var _expressAsyncHandler = _interopRequireDefault(require("express-async-handler"));

var _PaymailError = require("./errors/PaymailError");

var _httpStatusCodes = _interopRequireDefault(require("http-status-codes"));

var helpers = _interopRequireWildcard(require("./script-helpers"));

var _paymailClient = require("@moneybutton/paymail-client");

var validateRequest =
/*#__PURE__*/
function () {
  var _ref = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(params, paymailClient) {
    var message;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            if (params.signature) {
              _context.next = 2;
              break;
            }

            throw new _PaymailError.PaymailError('Missing signature', _httpStatusCodes.default.BAD_REQUEST, 'missing-signature');

          case 2:
            if (params.senderPaymail) {
              _context.next = 4;
              break;
            }

            throw new _PaymailError.PaymailError('Missing sender paymail', _httpStatusCodes.default.BAD_REQUEST, 'missing-sender-paymail');

          case 4:
            if (/^\S+@\S+.\S+$/.test(params.senderPaymail)) {
              _context.next = 6;
              break;
            }

            throw new _PaymailError.PaymailError('Invalid sender paymail', _httpStatusCodes.default.BAD_REQUEST, 'invalid-sender-paymail');

          case 6:
            if (params.dt) {
              _context.next = 8;
              break;
            }

            throw new _PaymailError.PaymailError('Missing parameter dt', _httpStatusCodes.default.BAD_REQUEST, 'missing-dt');

          case 8:
            message = _paymailClient.VerifiableMessage.forBasicAddressResolution({
              senderPaymail: params.senderPaymail,
              dt: params.dt,
              amount: params.amount,
              purpose: params.purpose
            });
            _context.next = 11;
            return paymailClient.isValidSignature(message, params.signature, params.senderPaymail, params.pubkey);

          case 11:
            if (_context.sent) {
              _context.next = 13;
              break;
            }

            throw new _PaymailError.PaymailError('Wrong signature', _httpStatusCodes.default.BAD_REQUEST, 'bad-signature');

          case 13:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function validateRequest(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var buildGetAddressRouter = function buildGetAddressRouter(config, ifPresent) {
  if (config.getPaymentDestination) {
    var router = _express.default.Router();

    router.post('/address/:paymail', (0, _expressAsyncHandler.default)(
    /*#__PURE__*/
    function () {
      var _ref2 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(req, res) {
        var _req$params$paymail$s, _req$params$paymail$s2, name, domain, output;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _req$params$paymail$s = req.params.paymail.split('@'), _req$params$paymail$s2 = (0, _slicedToArray2.default)(_req$params$paymail$s, 2), name = _req$params$paymail$s2[0], domain = _req$params$paymail$s2[1];
                _context2.next = 3;
                return validateRequest(req.body, config.paymailClient);

              case 3:
                _context2.next = 5;
                return config.getPaymentDestination(name, domain, req.body, helpers);

              case 5:
                output = _context2.sent;

                if (output) {
                  _context2.next = 8;
                  break;
                }

                throw new _PaymailError.PaymailError("Paymail not found: ".concat(req.params.paymail), _httpStatusCodes.default.NOT_FOUND, 'not-found');

              case 8:
                res.send({
                  output: output
                });

              case 9:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function (_x3, _x4) {
        return _ref2.apply(this, arguments);
      };
    }()));
    ifPresent(router);
  }
};

exports.buildGetAddressRouter = buildGetAddressRouter;
